File: src/main/java/leetCode/offer/Offer57_2.java
package leetCode.offer;

import java.util.ArrayList;
import java.util.List;

/**
 * @author jingxinwu
 * @date 2021-11-25 1:05 上午
 */
public class Offer57_2 {


    public int[][] findContinuousSequence(int target) {
        if (target <= 2) {
            return new int[][]{};
        }
        List<int[]> list = new ArrayList<>();
        for (int l = 1, r = 2; l < r; ) {
            int sum = (l + r) * (r - l + 1) / 2;
            if (sum == target) {
                int[] res = new int[r - l + 1];
                for (int i = l; i <= r; ++i) {
                    res[i - l] = i;
                }
                list.add(res);
                l++;
            } else if (sum < target) {
                r++;
            } else {
                l++;
            }

        }
        return list.toArray(new int[list.size()][]);
    }
}

File: src/main/go/leetcode/lists/offer/Offer57_2.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/3 20:11
 */

func findContinuousSequence(target int) [][]int {

	if target == 0 {
		return [][]int{}
	}
	n := target/2 + 1
	presum := make([]int, n+1)
	presum[0] = 0
	sum := 0
	for i := 1; i <= n; i++ {
		sum = sum + i
		presum[i] = sum
	}
	mp := make(map[int]int)
	res := make([][]int, 0)
	for i := 0; i < len(presum); i++ {
		if v, ok := mp[presum[i]-target]; ok {
			t := make([]int, 0)
			for j := v + 1; j <= i; j++ {
				t = append(t, j)
			}
			res = append(res, t)
		}
		mp[presum[i]] = i
	}
	return res
}
